### PR TITLE
JAMES-1776 Port GetMessagesMethodTest to Cucumber

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+    @Inject
+    private GetMessagesMethodStepdefs(MainStepdefs mainStepdefs, UserStepdefs userStepdefs) {
+        this.mainStepdefs = mainStepdefs;
+        this.userStepdefs = userStepdefs;
+    }
+
+    @Given("^the user has a message in \"([^\"]*)\" mailbox with subject \"([^\"]*)\" and content \"([^\"]*)\"$")
+    public void appendMessage(String mailbox, String subject, String content) throws Throwable {
+        appendMessage(mailbox, Optional.empty(), subject, content, Optional.empty());
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: please give a name to Optional.empty() for example : NO_CONTENT_TYPE

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    @Given("^the user has a message in \"([^\"]*)\" mailbox with subject \"([^\"]*)\" and content \"([^\"]*)\" with headers$")
+    public void appendMessage(String mailbox, String subject, String content, DataTable headers) throws Throwable {
+        appendMessage(mailbox, Optional.empty(), subject, content, Optional.of(headers.asMap(String.class, String.class)));
+    }
+
+    private void appendMessage(String mailbox, Optional<String> contentType, String subject, String content, Optional<Map<String, String>> headers) throws Exception {
+        ZonedDateTime dateTime = ZonedDateTime.parse("2014-10-30T14:12:00Z");
+        mainStepdefs.jmapServer.serverProbe().appendMessage(userStepdefs.username, 
+                new MailboxPath(MailboxConstants.USER_NAMESPACE, userStepdefs.username, mailbox),
+                new ByteArrayInputStream(message(contentType, subject, content, headers).getBytes()),
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: getBytes should always use a Charset

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    public void appendMessage(String mailbox, String contentType, String subject, String content) throws Throwable {
+        appendMessage(mailbox, Optional.of(contentType), subject, content, Optional.empty());
+    }
+
+    @Given("^the user has a message in \"([^\"]*)\" mailbox with subject \"([^\"]*)\" and content \"([^\"]*)\" with headers$")
+    public void appendMessage(String mailbox, String subject, String content, DataTable headers) throws Throwable {
+        appendMessage(mailbox, Optional.empty(), subject, content, Optional.of(headers.asMap(String.class, String.class)));
+    }
+
+    private void appendMessage(String mailbox, Optional<String> contentType, String subject, String content, Optional<Map<String, String>> headers) throws Exception {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: why not a real Content-Type object ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+                Date.from(dateTime.toInstant()), false, new Flags());
+    }
+
+    private String message(Optional<String> contentType, String subject, String content, Optional<Map<String,String>> headers) {
+        Function<Set<Map.Entry<String, String>>, String> entriesToString = entries -> entries.stream()
+                .map(this::entryToPair)
+                .map(this::joinKeyValue)
+                .collect(Collectors.joining("\r\n", "", "\r\n"));
+
+        String headersAsString = headers
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: should be another method

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+                .map(map -> map.entrySet())
+                .map(entriesToString)
+                .orElse("");
+                
+        String contentTypeAsString = contentType
+                .map(value -> "Content-Type: " + value + "\r\n")
+                .orElse("");
+
+        return headersAsString + contentTypeAsString + "Subject: " + subject + "\r\n\r\n" + content;
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: why don't we use mime4j ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        return Pair.with(String.format("\"%s\"", pair.getValue0()), pair.getValue1());
+    }
+
+    private String joinKeyValue(Pair<String, String> pair) {
+        return Joiner.on(": ").join(pair);
+    }
+
+    private void post(String requestBody) {
+        post = with()
+            .accept(ContentType.JSON)
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: You can configure defaults with : RestAssured.requestSpecification = blabla.contentType(blabla).accept(blabla)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 7]
```diff
@@ -0,0 +1,182 @@
+Feature: GetMessages method
+  As a James user
+  I want to be able to retrieve my messages
+
+  Background:
+    Given a domain named "domain.tld"
+    And a current user with username "username@domain.tld" and password "secret"
+    And the current user has a mailbox named "inbox"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: why do you have to repeat current user, it's already current ...

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+  Scenario: Retrieving messages with a non null accountId should return a NotSupported error
+    When the user is getting his messages and gives a non null accountId
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: When the user ask for "inbox" messages using its accountId

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    Then an error "Not yet implemented" is returned
+
+  Scenario: Unknown arguments should be ignored when retrieving messages
+    When the user is getting his messages and gives unknown arguments
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: When the user ask for "inbox" messages using unknown arguments

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    Then no error is returned
+    And the list of unknown messages is empty
+    And the list of messages is empty
+
+  Scenario: Retrieving messages with invalid argument should return an InvalidArguments error
+    When the user is getting his messages and gives invalid argument
+    Then an error "invalidArguments" is returned
+    And the description is "N/A (through reference chain: org.apache.james.jmap.model.Builder["ids"])"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: it should not ...

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+  Scenario: Retrieving messages should return empty list when no message
+    When the user is getting his messages
+    Then no error is returned
+    And the list of messages is empty
+
+  Scenario: Retrieving message should return not found when message dont exists
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: doesn't exist

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    When the user is getting his messages and gives a list of ids "["username@domain.tld|inbox|12"]"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: When the user ask for "inbox" messages X, Y, Z

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+    Then no error is returned
+    And the notFound list should contains "username@domain.tld|inbox|12"
+
+  Scenario: Retrieving message should return messages when exists
+    Given the user has a message in "inbox" mailbox with subject "my test subject" and content "testmail"
+    When the user is getting his messages and gives a list of ids "["username@domain.tld|inbox|1"]"
+    Then no error is returned
+    And the list should contain 1 message
+    And the id of the first message is "username@domain.tld|inbox|1"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: there's only one message, you don't have to repeat first message everywhere

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+    And the threadId of the first message is "username@domain.tld|inbox|1"
+    And the subject of the first message is "my test subject"
+    And the textBody of the first message is "testmail"
+    And the isUnread of the first message is "true"
+    And the preview of the first message is "testmail"
+    And the headers of the first message is map containing only
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: headers contains :

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+    Given the user has a message in <mailbox> mailbox with content-type <content-type> subject <subject> and content <content>
+    When the user is getting his messages and gives a list of ids "["username@domain.tld|inbox|1"]"
+    Then no error is returned
+    And the list should contain 1 message
+    And the id of the first message is "username@domain.tld|inbox|1"
+    And the threadId of the first message is "username@domain.tld|inbox|1"
+    And the subject of the first message is <subject>
+    And the htmlBody of the first message is <content>
+    And the isUnread of the first message is "true"
+    And the preview of the first message is <preview>
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: without quotes ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    And the headers of the first message is map containing only
+      |content-type |text/html |
+      |subject |<subject-header> |
+    And the date of the first message is "2014-10-30T14:12:00Z"
+
+    Examples:
+      |mailbox |content-type |subject |subject-header |content |preview |
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: you have to align table columns

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    And the subject of the first message is "my test subject"
+
+  Scenario: Retrieving message should return attachments when some
+    Given the user has a message in "inbox" mailbox with two attachments
+    When the user is getting his messages and gives a list of ids "["username@domain.tld|inbox|1"]"
+    Then no error is returned
+    And the list should contain 1 message
+    And the hasAttachment of the first message is "true"
+    And the list of attachments of the first message contains 2 attachments
+    And the first attachment blobId is "223a76c0e8c1b1762487d8e0598bd88497d73ef2"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: why don't you use a table to describe each message ?

**Comment by aduprat** [comment]
test this please

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java (hunk 5 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: Then, we should modify then appendMesssage method in order to take a mime4j Message in replacement of all parameters.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java (hunk 6 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 7 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java (hunk 5 above)
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessagesMethodStepdefs.java]: I don't see why. We can very well use mime4j to generate some bytes without breaking everything

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 8 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: Ok, but without "inbox" as fetching in a dedicated mailbox is not in the GetMessages specification.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 9 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 10 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: That's current behaviour, this issue is just about test migration.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 11 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 12 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 13 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 14 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 10 above)
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: but we should focus on our gherkin quality : there's no way this description could be correct, if you think we don't have time to fix it, fix the test anyway and mark it as ignored

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 16 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 17 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 10 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: https://issues.apache.org/jira/browse/JAMES-1782

**Comment by aduprat** [comment]
#300

**Pull request closed without merging**